include/linker/devicetree_reserved.h

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2021, Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Generate memory regions and sections from reserved-memory nodes.
+ */
+
+/* Reserved regions under chosen/zephyr,reserved */
+#define _CHOSEN_RESERVED DT_CHOSEN(zephyr_reserved)
+
+/* Unquoted region label */
+#define _DT_LABEL_TOKEN(res) DT_STRING_TOKEN(res, label)
+
+/* _start and _end section symbols */
+#define _DT_RESERVED_PREFIX(res) UTIL_CAT(reserved_, _DT_LABEL_TOKEN(res))
+#define _DT_RESERVED_START(res)  UTIL_CAT(_DT_RESERVED_PREFIX(res), _start)
+#define _DT_RESERVED_END(res)    UTIL_CAT(_DT_RESERVED_PREFIX(res), _end)
+
+/* Declare a reserved memory region */
+#define _RESERVED_REGION_DECLARE(res) DT_STRING_TOKEN(res, label) (rw) :	\
+				       ORIGIN = DT_REG_ADDR(res),		\
+				       LENGTH = DT_REG_SIZE(res)
+
+/* Declare a reserved memory section */
+#define _RESERVED_SECTION_DECLARE(res) _DT_LABEL_TOKEN(res) :			\
+					{					\
+					  _DT_RESERVED_START(res) = .;		\
+					  KEEP(*(._DT_LABEL_TOKEN(res)))	\
+					  KEEP(*(._DT_LABEL_TOKEN(res).*))	\
+					  . += DT_REG_SIZE(res);		\
+					  _DT_RESERVED_END(res) = .;		\
+					} > _DT_LABEL_TOKEN(res)
```

Comment:
@tejlmand will need to review this to see how it's going to play with linker script generation, which is also a 2.7 feature.

@carlocaione my expectation is that you may need to lift this to cmake instead of .h at some point, based on what @tejlmand demonstrated at the TSC meeting today regarding toolchain abstractions and impact on linker scripts.

I'll let y'all sort out the details; I'm just calling attention to it.

Reply:
I was not aware of @tejlmand work, thanks for pointing that out. Looking forward to @tejlmand comments then ;)

Reply:
Indeed, macro generation of linker sections is going to clash directly with @tejlmand's work. But since this RFC is probably going to come in earlier than the linker script generation rework, we should make sure that @tejlmand takes this into account during his development.

Reply:
I'm not so worried on this.
It would of require that i'm able to do similar, but as we anyway need to obtain device tree info for the linker script generation I don't have concerns in this regard.